docs/proposals/20190708-MachineConfig.md

```diff
@@ -0,0 +1,293 @@
+# Proposal: Machine Configuration in Talos v0.2.0 and Beyond
+
+Author(s): Spencer Smith (@rsmitty)
+
+## Abstract
+
+This proposal will outline how we'll handle the passing of machine configurations in Talos, as we're aiming to simplify the options and keep the user from having to worry about the minutiae of kubeadm and all of the options.
+
+## Background
+
+I think the easiest way to background this is to take a look at each of the machine configs that we currently have templates for. When looking at these files, it's somewhat self-explanatory on what is available to tweak, but it also gives a good idea of what is similar between the three types of Talos nodes: init (the first master), control plane (any other masters), and workers. I've also appended some additional fields that we use for certain platforms like Packet.
+
+#### Init:
+```
+#!talos
+version: ""
+security:
+  os:
+    ca:
+      crt: "{{ .Certs.OsCert }}"
+      key: "{{ .Certs.OsKey }}"
+  kubernetes:
+    ca:
+      crt: "{{ .Certs.K8sCert }}"
+      key: "{{ .Certs.K8sKey }}"
+services:
+  init:
+    cni: flannel
+  kubeadm:
+    initToken: {{ .InitToken }}
```

Comment:
Should this even be exposed at all?  This is quite ephemeral (default 24-hours), and doesn't have anything to do with the PKI infrastructure (so far as I know).

Reply:
Since we have no mechanism to tell `kubeadm` about any new tokens, we actually disable the TTL and make the token never expire. Happy to revisit. The thing that makes expiring `kubeadm` tokens difficult is automation. 🤔 I wonder if we can set tokens on the fly with CAPI.

Reply:
Ultimately, we'll need to revisit the perpetual tokens.  This should be something that trustd can handle:  generating new kubeadm tokens to add new machines (which can be done by the kubeadm CLI tool, so it really _should_ be an API).

For now, though, if it makes it easier, it doesn't really hurt anything.

Reply:
100% Agree... this should absolutely be an API. So would it be something like the trust token is the single long standing token and then by virtue of the fact that the trustd token is valid, that a machine can request a token?

Reply:
Pretty much, yes.

Down the road, of course, I'd still like even the trustd token to at least have an _optional_ expiry and provide a facility via osctl to generate new ones (to be used in future machine configs).  This is why I wanted to be able to bind in and register machine identifiers, so that we could have ephemeral trustd tokens:  once a machine is registered, its own machine ID acts as a token.  New machines require new tokens, within a deployment window.  That may not work well for cloud environments with AutoScaling (which is why it should be optional), but it's a very nice optional security level.

Reply:
Agree.